File: ios/SharedList/Features/Auth/PreviewSupport.swift
import Foundation
import SwiftData

#if DEBUG
// Helpers used only by SwiftUI #Preview blocks. Compiled out of release
// builds via #if DEBUG so the production binary doesn't carry preview-only
// code paths.
//
// We construct an AppContainer with an in-memory Keychain (real Keychain
// behavior in previews is unreliable across Xcode versions) and a stub
// AuthService that doesn't hit the network. Previews that just need to
// render UI in different states (logged in / logged out) can use these
// without standing up the live backend.

@MainActor
enum PreviewSupport {
    static func loggedOutContainer() -> AppContainer {
        makeContainer(behavior: .alwaysLoggedOut, seedSession: false)
    }

    static func loggedInContainer() -> AppContainer {
        // Seed a session synchronously so the preview comes up with state.
        // The Task is fire-and-forget; the preview view re-renders when the
        // state lands, which is exactly the behavior we'd see at runtime.
        let container = makeContainer(behavior: .alwaysLoggedIn, seedSession: true)
        return container
    }

    private static func makeContainer(behavior: StubAuthService.Behavior, seedSession: Bool) -> AppContainer {
        let keychain = InMemoryKeychainStore()
        let tokenStore = TokenStore(keychain: keychain)
        let api = APIClient(baseURL: AppContainer.defaultBaseURL, tokenStore: tokenStore)
        let auth = StubAuthService(tokenStore: tokenStore, behavior: behavior)
        let monitor = MockNetworkMonitor(isOnline: true)
        let modelContainer = previewModelContainer()
        let syncEngine = SyncEngine(
            api: api,
            container: modelContainer,
            monitor: monitor,
            currentUserId: { [weak auth] in auth?.currentUser()?.id }
        )
        if seedSession {
            Task { @MainActor in
                try? await tokenStore.save(.init(
                    accessToken: "preview-access",
                    refreshToken: "preview-refresh",
                    user: AuthUser(id: "preview-user", email: "alice@example.com", displayName: "Alice")
                ))
            }
        }
        return AppContainer(
            keychain: keychain,
            tokenStore: tokenStore,
            api: api,
            auth: auth,
            networkMonitor: monitor,
            modelContainer: modelContainer,
            syncEngine: syncEngine
        )
    }

    /// Build a SwiftData container with `isStoredInMemoryOnly: true` so each
    /// preview / test session gets a fresh empty store and writes don't bleed
    /// across runs. The schema lists every `@Model` type the app uses; if a
    /// type is added in production code, this list must grow too.
    static func previewModelContainer() -> ModelContainer {
        let configuration = ModelConfiguration(isStoredInMemoryOnly: true)
        do {
            return try ModelContainer(
                for: UserModel.self,
                ListModel.self,
                ItemModel.self,
                MemberModel.self,
                SyncCursor.self,
                configurations: configuration
            )
        } catch {
            fatalError("Failed to construct preview ModelContainer: \(error)")
        }
    }
}

@MainActor
final class StubAuthService: AuthServicing {
    enum Behavior {
        case alwaysLoggedOut
        case alwaysLoggedIn
    }

    private let tokenStore: TokenStore
    private let behavior: Behavior

    init(tokenStore: TokenStore, behavior: Behavior) {
        self.tokenStore = tokenStore
        self.behavior = behavior
    }

    func signup(email: String, password: String, displayName: String) async throws -> AuthUser {
        let user = AuthUser(id: "stub-id", email: email, displayName: displayName)
        if behavior == .alwaysLoggedIn {
            try await tokenStore.save(.init(
                accessToken: "stub-access",
                refreshToken: "stub-refresh",
                user: user
            ))
        }
        return user
    }

    func login(email: String, password: String) async throws -> AuthUser {
        let user = AuthUser(id: "stub-id", email: email, displayName: "Stub User")
        if behavior == .alwaysLoggedIn {
            try await tokenStore.save(.init(
                accessToken: "stub-access",
                refreshToken: "stub-refresh",
                user: user
            ))
        }
        return user
    }

    func logout() async {
        await tokenStore.clear()
    }

    func currentUser() -> AuthUser? {
        tokenStore.current?.user
    }
}
#endif

File: android/app/src/main/java/in/santosh_bharadwaj/sharedlist/features/auth/PreviewSupport.kt
package `in`.santosh_bharadwaj.sharedlist.features.auth

import `in`.santosh_bharadwaj.sharedlist.app.AppContainer
import `in`.santosh_bharadwaj.sharedlist.core.auth.AuthService
import `in`.santosh_bharadwaj.sharedlist.core.auth.AuthUser
import `in`.santosh_bharadwaj.sharedlist.core.auth.TokenStore
import `in`.santosh_bharadwaj.sharedlist.core.networking.ApiClient
import `in`.santosh_bharadwaj.sharedlist.core.storage.InMemorySecureStorage
import kotlinx.coroutines.runBlocking

/**
 * Helpers used only by `@Preview` composables. Mirrors iOS PreviewSupport.
 *
 * We construct an AppContainer with [InMemorySecureStorage] (real
 * EncryptedSharedPreferences requires a Context that the Compose preview
 * runner doesn't always supply cleanly) and a stub [AuthService] that doesn't
 * hit the network. Previews that just need to render UI in different states
 * (logged in / logged out) use these without standing up the live backend.
 *
 * In production code a previews-only file like this would be `src/debug/`-only
 * to avoid shipping the stubs in release. For now the cost is a few unused
 * KB; we'll move this if the binary size ever matters.
 */
internal object PreviewSupport {

    fun loggedOutContainer(): AppContainer {
        val storage = InMemorySecureStorage()
        val tokenStore = TokenStore(storage)
        // ApiClient needs SOME engine, but StubAuthService never invokes it
        // in preview rendering. The default OkHttp engine is fine; it doesn't
        // open a real socket until a request actually fires.
        val api = ApiClient(baseUrl = AppContainer.DEFAULT_BASE_URL, tokenStore = tokenStore)
        val auth = StubAuthService(tokenStore = tokenStore, behavior = StubAuthService.Behavior.AlwaysLoggedOut)
        return AppContainer.forTesting(storage, tokenStore, api, auth)
    }

    fun loggedInContainer(): AppContainer {
        val storage = InMemorySecureStorage()
        val tokenStore = TokenStore(storage)
        val api = ApiClient(baseUrl = AppContainer.DEFAULT_BASE_URL, tokenStore = tokenStore)
        val auth = StubAuthService(tokenStore = tokenStore, behavior = StubAuthService.Behavior.AlwaysLoggedIn)
        // Seed a session so the preview comes up with state. runBlocking is
        // used here because preview execution is synchronous and short-lived;
        // in production code we'd never block a coroutine context like this.
        runBlocking {
            tokenStore.save(
                TokenStore.Tokens(
                    accessToken = "preview-access",
                    refreshToken = "preview-refresh",
                    user = AuthUser(id = "preview-user", email = "alice@example.com", displayName = "Alice"),
                ),
            )
        }
        return AppContainer.forTesting(storage, tokenStore, api, auth)
    }
}

internal class StubAuthService(
    private val tokenStore: TokenStore,
    private val behavior: Behavior,
) : AuthService {
    enum class Behavior { AlwaysLoggedOut, AlwaysLoggedIn }

    override suspend fun signup(email: String, password: String, displayName: String): AuthUser {
        val user = AuthUser(id = "stub-id", email = email, displayName = displayName)
        if (behavior == Behavior.AlwaysLoggedIn) {
            tokenStore.save(
                TokenStore.Tokens(
                    accessToken = "stub-access",
                    refreshToken = "stub-refresh",
                    user = user,
                ),
            )
        }
        return user
    }

    override suspend fun login(email: String, password: String): AuthUser {
        val user = AuthUser(id = "stub-id", email = email, displayName = "Stub User")
        if (behavior == Behavior.AlwaysLoggedIn) {
            tokenStore.save(
                TokenStore.Tokens(
                    accessToken = "stub-access",
                    refreshToken = "stub-refresh",
                    user = user,
                ),
            )
        }
        return user
    }

    override suspend fun logout() {
        tokenStore.clear()
    }

    override fun currentUser(): AuthUser? = tokenStore.current?.user
}
